Extends call/fetchRemote

xLuxy opened the pull request:
Fixes #370

This Pr allows more control over fetch/callRemote

Adds the following functions (Server only for feedback)
`table getNetworkRequests( [string/resource resource] )`
`table/false getNetworkRequestInfo(userdata requestHandler)`

`getNetworkRequestInfo` returns a table containing the following info(example dump):
```json
[  
   {  
      "post_data":"[]",
      "resource":"^R^zombies",
      "queue":"default",
      "type":"call",
      "connection_timeout":10000,
      "connection_attempts":10,
      "start":1539790689,
      "url":"http:\/\/192.168.178.24:22022?v=C4CA4238A0B923820DCC509A6F75849B",
      "method":"POST",
      "headers":{  
         "Content-Type":"application\/json"
      }
   }
]
```

Example script used for testing:
```lua
function printTable(tblTable)
    for i, v in pairs(tblTable) do
        if(type(v) == "table") then
            outputServerLog("Printing table " .. i .. ':')
            outputServerLog("-------------------------------")
            printTable(v)
            outputServerLog("-------------------------------")
        else
            outputServerLog(("%s(%s) -> %s(%s)"):format(tostring(i), type(i), tostring(v), type(v)))
        end
    end
end

function dummyFunction() end

for i = 1, 3 do
    local uRemoteCall = callRemote("http://192.168.178.24:22022?v=" .. md5(i), dummyFunction)
    
    outputServerLog("Remotecall created: " .. tostring(uRemoteCall))
end

for _, uRemoteCall in ipairs(getNetworkRequests(resourceRoot)) do
    printTable(getNetworkRequestInfo(uRemoteCall))
end

local uVehicle = createVehicle(411, 0, 0, 0)

outputServerLog("Valid request test #1: " .. tostring(getNetworkRequestInfo("21212")))
outputServerLog("Valid request test #2: " .. tostring(getNetworkRequestInfo(true)))
outputServerLog("Valid request test #3: " .. tostring(getNetworkRequestInfo(resourceRoot)))
outputServerLog("Valid request test #4: " .. tostring(getNetworkRequestInfo(resourceRoot)))
outputServerLog("Valid request test #5: " .. tostring(getNetworkRequestInfo(uVehicle)))


-- testing download info
function dummyCallback2(strResponse, iErrorCode)
    outputServerLog("Response: " .. iErrorCode)
end

local uRemoteCall = fetchRemote("https://speed.hetzner.de/10GB.bin", dummyCallback2)
local uTimer      = false

outputServerLog("Download test: " .. tostring(uRemoteCall))

function checkDownload()
    local tblCallInfo = getNetworkRequestInfo(uRemoteCall)
    
    if(not tblCallInfo) then
        outputServerLog("Call was aborted / completed")
        killTimer(uTimer)
        
        uTimer = false
        return
    end
    
    local iBytesReceived = tblCallInfo.bytes_received
    local iTotalBytes    = tblCallInfo.bytes_total
    
    outputServerLog(("Download progress: %d / %d (%.02f)"):format(iBytesReceived, iTotalBytes, (iBytesReceived / iTotalBytes) * 100))
end

if(uRemoteCall) then
    uTimer = setTimer(checkDownload, 1000, 0)
end
```

Feedback is highly welcomed. More informations can be found here: https://github.com/multitheftauto/mtasa-blue/issues/612

- Luxy.c

xLuxy committed:
- Add table getNetworkRequests([string/resource resource])
- Add table/false getNetworkRequestInfo(userdata requestHandler)
- Made fetch/callRemote return userdata of CRemoteCall

xLuxy committed:
- Add table getNetworkRequests([string/resource resource])
- Add table/false getNetworkRequestInfo(userdata requestHandler)
- Made fetch/callRemote return userdata of CRemoteCall

xLuxy committed:
Merge branch 'enhancements/callFetchRemote' of https://github.com/xLuxy/mtasa-blue into enhancements/callFetchRemote

CrosRoad95 commented:
you should also introduce, how many attempts left

Commit:
Merge branch 'master' into enhancements/callFetchRemote

xLuxy committed:
Fix line spacing

xLuxy committed:
Fixed errors from previous commit

xLuxy committed:
Add clientsided getNetworkRequests & getNetworkRequestInfo

ghost commented:
Added clientsided stuff - `abortNetworkRequest` is next

xLuxy committed:
Add shared abortNetworkRequest

ccw808 commented on Client/mods/deathmatch/logic/CRemoteCalls.cpp:
```diff
@@ -219,6 +248,13 @@ CRemoteCall::~CRemoteCall()
+    time(&timer);
+    localtime(&timer);
+    
+    m_iStartTime = timer;
```
Debug code?

ccw808 commented on Client/mods/deathmatch/logic/CRemoteCalls.cpp:
```diff
@@ -28,34 +29,62 @@ CRemoteCalls::~CRemoteCalls()
-    m_calls.back()->MakeCall();
+    CRemoteCall* pRemoteCall = new CRemoteCall(szURL, arguments, luaMain, iFunction, strQueueName, uiConnectionAttempts, uiConnectTimeoutMs);
+
+    if (pRemoteCall)
```
I don't think `new` will return nullptr

botder replied:
Note: `new` throws `std::bad_alloc`, but you can select the non-throwing allocation function by using `std::nothrow`

botder committed:
Merge branch 'master'

xLuxy committed:
Added CallExists checks to CLuaFunctionParseHelpers::UserDataCast

xLuxy committed:
Removed unnessesary checks

xLuxy committed:
WIP: Added bytes_received / bytes_total to getNetworkRequestInfo

ghost commented:
bytes_received / bytes total seems to be broken @ccw808 
also, fetchRemote seems to be broken - will investigate
also updated example script used above

- xLuxy

xLuxy committed:
Merge branch 'master' into enhancements/callFetchRemote

xLuxy committed:
Added missing return in CallRemote

ccw808 commented:
I'm wondering if some of the info returned by `getNetworkRequestInfo` should be removed for security reasons. Otherwise any resource could monitor remote calls for sensitive data in the URL, post data or headers.

CrosRoad95 commented:
> should be removed

if resource don't have persmission or was send from other resource

ghost commented:
Well, each server owner is responsible to check their running resources for
security issues. PostData might be useful for debugging same as headers. We
might want to introduce another acl permission for that.

Am Fr., 11. Jan. 2019, 14:00 hat ccw <notifications@github.com> geschrieben:

> I'm wondering if some of the info returned by getNetworkRequestInfo
> should be removed for security reasons. Otherwise any resource could
> monitor remote calls for sensitive data in the URL, post data or headers.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/multitheftauto/mtasa-blue/pull/660#issuecomment-453509655>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHqwncN4_3akv90LDizTmIPgqq_YStqPks5vCIr8gaJpZM4XkR9b>
> .
>

botder commented:
We could also use the ACL to manage the access to that information, but allow access by the resource, which made the call to callRemote/fetchRemote in the first place.

ghost commented:
What about these acl rights:

function.getNetworkRequestInfo (default: false)
general.advancedRequestInfo (default: false)

In general, all resources will need access to function.getNetworkRequestInfo. The Resource who had called fetchRemote doesnt need the general.advancedRequestInfo permission.

general.advancedRequestInfo is basically to hide post data & headers. 

What do you guys think?

// Edit: should getNetworkRequests also have an acl permission?

xLuxy committed:
Merge branch 'master' into enhancements/callFetchRemote

xLuxy committed:
Testfix for "invalid use of incomplete type" CRemoteCalls

xLuxy committed:
Testfix 2 for "invalid use of incomplete type" CRemoteCalls

xLuxy committed:
Add ACL permission check for getNetworkRequestInfo

xLuxy committed:
Remove Bad argument @ message from getNetworkRequestInfo

xLuxy committed:
Addendum to previous commit - also removed clientsided debug message

xLuxy committed:
Only allow getNetworkRequestInfo for its own resource clientsided

Commit:
Merge branch 'master' into enhancements/callFetchRemote

ghost commented:
@ccw808: Hope its secure enough now and ready to be merged.

Test resource used while testing: http://149.202.86.45/mta_fetchCallRemoteEnhancement.zip

- Luxy.c

ccw808 commented:
I think we should change `getNetworkRequestInfo` to avoid confusion with unrelated `getNetworkUsageData` and `getNetworkStats`

Suggestions:
- A) `getHttpRequestInfo`
- B) `getRemoteRequestInfo`
- C) `getFetchRemoteRequestInfo`
- D) `getFetchRemoteInfo`

or something else?

CrosRoad95 commented:
C

ghost commented:
What do you guys think about the following:

`getCallRemoteRequests`, `getFetchRemoteRequests`
`getCallRemoteRequestInfo`, `getFetchRemoteRequestInfo`
`abortHTTPRequest`

would that be better ?

qaisjp commented:
What about `getHTTPRequestInfo`?

ghost commented:
> What about `getHTTPRequestInfo`?


This would probably not be good aswell as there are http functions like https://wiki.multitheftauto.com/wiki/HttpWrite, https://wiki.multitheftauto.com/wiki/HttpRequestLogin etc

xLuxy committed:
Rename getNetworkRequests to getRemoteRequests
Rename getNetworkRequestInfo to getRemoteRequestInfo
Rename abortNetworkRequest to abortRemoteRequest

xLuxy committed:
Renamed bytes_received -> bytesReceives
Renamed bytes_total -> bytesTotal
Added currentAttempt (0 = queued, 1+ = Downloading)

xLuxy committed:
Renamed post_data -> postData
Renamed connection_(attempts/total) -> connection(Attempts/Total)

xLuxy committed:
Remove RemoteCall when aborted

ghost commented:
I think that they're fine now, updated test resources can be downloaded here: http://149.202.86.45/mta_fetchCallRemoteEnhancement.zip

xLuxy committed:
Merge remote-tracking branch 'origin/master' into enhancements/callFetchRemote

ccw808 commented:
What about having no restrictions on getting basic info, and have a special ACL right of `general.fullRemoteRequestInfo` to get extended info. i.e. post fields+headers+url args

xLuxy committed:
Fixed luaVM not being used in GetRemoteRequests

ghost commented:
Heres the test resource used for testing: [fr.zip](https://github.com/multitheftauto/mtasa-blue/files/3169523/fr.zip)

@qaisjp @ccw808

qaisjp commented:
> What about having no restrictions on getting basic info, and have a special ACL right of `general.fullRemoteRequestInfo` to get extended info. i.e. post fields+headers+url args

What about also automatically granting full request info for the source resource of the request?

xLuxy committed:
Cleanup Serversided stuff

xLuxy committed:
Fixed some more serversided stuff and cleanup clientside

ghost commented:
So, for now, the functions should be as follow:
```lua
table getRemoteRequestInfo(element theRequest)
bool abortRemoteRequest(element theRequest)
table getRemoteRequests([resource theResource = nil])
```

- Luxy

xLuxy committed:
Merge branch 'master' into enhancements/callFetchRemote

ccw808 commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1920,3 +1922,139 @@ int CLuaFunctionDefs::FetchRemote(lua_State* luaVM)
+
+        // TODO: remove query_string from url when bExtendedInfo isn't set
+        info.PushString("url");
+        info.PushString(pRemoteCall->GetURL().c_str());
```
Something like this to to remove query string? `pRemoteCall->GetURL().ReplaceI("%3F","?").Replace("#","?").SplitLeft("?");`

ccw808 commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1920,3 +1922,139 @@ int CLuaFunctionDefs::FetchRemote(lua_State* luaVM)
+        if (bExtendedInfo)
+        {
+            info.PushString("postData");
+            info.PushString(pRemoteCall->GetOptions().strPostData.c_str());
```
The post data could be very large. Suggest we have an optional parameter:
`getRemoteRequestInfo(element theRequest, int maxPostDataSize=0)`

ccw808 commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1920,3 +1922,139 @@ int CLuaFunctionDefs::FetchRemote(lua_State* luaVM)
+            info.PushBoolean(false);
+
+        info.PushString("start");
+        info.PushNumber(pRemoteCall->GetStartTime());
```
I think we should return the number of ticks since the download was queued

ccw808 commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1920,3 +1922,139 @@ int CLuaFunctionDefs::FetchRemote(lua_State* luaVM)
+            info.PushNumber(downloadInfo.uiContentLength);
+        }
+
+        info.PushString("currentAttempt");
```
Should this be inside previous `if{ ... }` block?

ccw808 replied:
Yes, I was talking about `currentAttempt`.
However, after looking at `CRemoteCall::GetDownloadStatus()` I realized that returning false can be avoided, so it has been refactored to return `const SDownloadStatus&` instead. You'll have to update this section of code to cope with that change.

xLuxy committed:
Merge branch 'master' into enhancements/callFetchRemote

xLuxy committed:
Merge branch 'master' into 'enhancements/callFetchRemote'

xLuxy committed:
Resolve merge errors introduced in e77c8cd0af4eb1accce74c843f89dc793f4287c4

xLuxy committed:
Revert 'Server/mods/deathmatch/logic/CRemoteCalls.cpp' to 284bf73cb2d0431ba3e7b38055698bb389b700d3

xLuxy committed:
Use GetTickCount64_() instead of current unix timestamp

xLuxy committed:
Remove query_string parameters from URL

xLuxy committed:
Add postDataLength parameter

ghost commented:
Done @ccw808

ccw808 commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1920,3 +1922,147 @@ int CLuaFunctionDefs::FetchRemote(lua_State* luaVM)
+            SString sPostData = pRemoteCall->GetOptions().strPostData.c_str();
+
+            if (iPostDataLength > 0)
+                sPostData = sPostData.SubStr(0, iPostDataLength);
```
If `iPostDataLength` is 0, then all the post data is included. Is that the intention?

ccw808 replied:
The problem I have with including postdata by default (and headers to a lesser extent) is that they take memory and cpu, don't change during the download and might not be required by scripters

sbx320 replied:
Defaulting to 100 is weird and will cause issues with scripters only testing things for low amounts of post data. I think a better option would be to expect an option table to include the more resource intensive data on-demand only. 

`getRemoteRequestInfo(..., { postDataLength = 0 })` -> Infinite length
`getRemoteRequestInfo(...)` -> No post data
`getRemoteRequestInfo(..., { postDataLength = 500 })` -> Limited to 500.

Same thing with headers:

`getRemoteRequestInfo(..., { includeHeaders = true })` -> Adds headers

sbx320 commented on Client/mods/deathmatch/logic/lua/CLuaFunctionDefs.World.cpp:
```diff
@@ -1920,3 +1922,147 @@ int CLuaFunctionDefs::FetchRemote(lua_State* luaVM)
+
+        if (bExtendedInfo)
+        {
+            SString sPostData = pRemoteCall->GetOptions().strPostData.c_str();
```
I think this may cause issues with binary post data. Is there a specific reason why you use `c_str()`  here?

xLuxy committed:
Add includeHeaders parameter, clean up few lines of code and also modified postDataLenght parameter to either return EVERYTHING (-1) or a specific length (> 0)
